### PR TITLE
[SPARK-30184][SQL] Implement a helper method for aliasing functions

### PR DESCRIPTION
**Pull request by amanomer**
### What changes were proposed in this pull request?
This PR is to use `expressionWithAlias` for remaining functions for which alias name can be used. Remaining functions are:
`Average, First, Last, ApproximatePercentile, StddevSamp, VarianceSamp`

PR https://github.com/apache/spark/pull/26712 introduced `expressionWithAlias`
### Why are the changes needed?
Error message is wrong when alias name is used for above mentioned functions.
### Does this PR introduce any user-facing change?
No
### How was this patch tested?
Manually

### CODE REVIEW

**Comment by amanomer** [comment]
cc @cloud-fan

**Comment by amanomer** [comment]
cc @cloud-fan

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala [hunk 1]
```diff
@@ -198,7 +198,8 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
 
         // Select the result of the first aggregate in the last aggregate.
         val result = AggregateExpression(
-          aggregate.First(evalWithinGroup(regularGroupId, operator.toAttribute), Literal(true)),
+          aggregate.First("first",
+            evalWithinGroup(regularGroupId, operator.toAttribute), Literal(true)),
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala]: a bit verbose to always set "first" for `First`, so how about defining auxiliary constructor?
```
  def this(child: Expression, ignoreNullsExpr: Expression) =
    this("first", child, Literal.create(false, BooleanType))
```

**Review by maropu — APPROVED** [review]
If the tests pass, LGTM except for one minor comment

**Comment by maropu** [comment]
btw, why did you include fixes for Average and ApproximatePercentile in this pr?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 2]
```diff
@@ -302,14 +302,14 @@ object FunctionRegistry {
     expression[Skewness]("skewness"),
     expression[ApproximatePercentile]("percentile_approx"),
     expression[ApproximatePercentile]("approx_percentile"),
-    expression[StddevSamp]("std"),
-    expression[StddevSamp]("stddev"),
+    expressionWithAlias[StddevSamp]("std"),
+    expressionWithAlias[StddevSamp]("stddev"),
     expression[StddevPop]("stddev_pop"),
-    expression[StddevSamp]("stddev_samp"),
+    expressionWithAlias[StddevSamp]("stddev_samp"),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: can we put them closer if they use the same expression?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 3]
```diff
@@ -431,7 +431,7 @@ object functions {
    * @since 2.0.0
    */
   def first(e: Column, ignoreNulls: Boolean): Column = withAggregateFunction {
-    new First(e.expr, Literal(ignoreNulls))
+    new First("first", e.expr, Literal(ignoreNulls))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: We can add a constructor to provide a default name.

**Comment by amanomer** [comment]
> btw, why did you include fixes for Average and ApproximatePercentile in this pr?

`Average` and `ApproximatePercentile` can be called with alias name as well.
https://github.com/apache/spark/blob/be867e8a9ee8fc5e4831521770f51793e9265550/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala#L303-L304

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 2 above)
Comment (amanomer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Sure

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 3 above)
Comment (amanomer) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala (hunk 1 above)
Comment (amanomer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala [hunk 4]
```diff
@@ -198,7 +198,7 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
 
         // Select the result of the first aggregate in the last aggregate.
         val result = AggregateExpression(
-          aggregate.First(evalWithinGroup(regularGroupId, operator.toAttribute), Literal(true)),
+          new aggregate.First(evalWithinGroup(regularGroupId, operator.toAttribute), Literal(true)),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala]: so adding more constructors doesn't help us to reduce diff, how about
```
case class First(funcName: String, child: Expression, ignoreNullsExpr: Expression) {
  def this(funcName: String, child: Expression)
}
object First {
  def apply(child: Expression, ignoreNullsExpr: Expression) ...
  def apply(child: Expression) ...
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 5]
```diff
@@ -126,6 +126,7 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession {
     //   1. Maven can't get correct resource directory when resources in other jars.
     //   2. We test subclasses in the hive-thriftserver module.
     val sparkHome = {
+      sys.props.put("spark.test.home", "/home/root1/spark")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: ?

**Review by maropu — APPROVED** [review]
LGTM except for one comment.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala (hunk 5 above)
Comment (amanomer) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: opps...I'll revert this

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala (hunk 4 above)
Comment (amanomer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala]: Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/LastTestSuite.scala [hunk 6]
```diff
@@ -24,7 +24,8 @@ import org.apache.spark.sql.types.IntegerType
 class LastTestSuite extends SparkFunSuite {
   val input = AttributeReference("input", IntegerType, nullable = true)()
   val evaluator = DeclarativeAggregateEvaluator(Last(input, Literal(false)), Seq(input))
-  val evaluatorIgnoreNulls = DeclarativeAggregateEvaluator(Last(input, Literal(true)), Seq(input))
+  val evaluatorIgnoreNulls = DeclarativeAggregateEvaluator(
+    Last(input, Literal(true)), Seq(input))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/LastTestSuite.scala]: nit: unnecessary change

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 7]
```diff
@@ -794,12 +794,18 @@ class ExpressionParserSuite extends AnalysisTest {
...
 
   test("Support respect nulls keywords for first_value and last_value") {
-    assertEqual("first_value(a ignore nulls)", First('a, Literal(true)).toAggregateExpression())
-    assertEqual("first_value(a respect nulls)", First('a, Literal(false)).toAggregateExpression())
-    assertEqual("first_value(a)", First('a, Literal(false)).toAggregateExpression())
-    assertEqual("last_value(a ignore nulls)", Last('a, Literal(true)).toAggregateExpression())
-    assertEqual("last_value(a respect nulls)", Last('a, Literal(false)).toAggregateExpression())
-    assertEqual("last_value(a)", Last('a, Literal(false)).toAggregateExpression())
+    assertEqual("first_value(a ignore nulls)",
+      First('a, Literal(true)).toAggregateExpression())
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: ditto, unnecessary change

File: sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala [hunk 8]
```diff
@@ -254,7 +254,8 @@ object StatFunctions extends Logging {
         stats.toLowerCase(Locale.ROOT) match {
           case "count" => (child: Expression) => Count(child).toAggregateExpression()
           case "mean" => (child: Expression) => Average(child).toAggregateExpression()
-          case "stddev" => (child: Expression) => StddevSamp(child).toAggregateExpression()
+          case "stddev" => (child: Expression) =>
+            StddevSamp("stddev_samp", child).toAggregateExpression()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala]: can we create a `object StddevSamp` and define the default name in the `apply` method?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala]: shall we use `stddev` here?

**Review by cloud-fan — APPROVED** [review]
LGTM except a few minor comments

**Comment by gengliangwang** [comment]
retest this please.

File: sql/core/src/test/resources/sql-tests/results/udf/postgreSQL/udf-aggregates_part1.sql.out [hunk 9]
```diff
@@ -101,7 +101,7 @@ struct<var_pop(CAST(CAST(udf(cast(b as string)) AS FLOAT) AS DOUBLE)):double>
 -- !query 12
 SELECT udf(var_samp(b)) FROM aggtest
 -- !query 12 schema
-struct<CAST(udf(cast(var_samp(cast(b as double)) as string)) AS DOUBLE):double>
+struct<CAST(udf(cast(var_samp(var_samp, cast(b as double)) as string)) AS DOUBLE):double>
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/udf/postgreSQL/udf-aggregates_part1.sql.out]: This looks weird. Looking at `Expression.toString`, I think we should override `flatArguments` in these expressions with alias to exclude the funcName parameter.

**Comment by amanomer** [comment]
Other class (`First`, `Last`) which takes `funcName` as arguments, I think we should override `flatArguments` there also?
Currently, no test cases will fail because of it but in future this may create problem.
cc @cloud-fan

File: sql/core/src/test/resources/sql-tests/results/udf/postgreSQL/udf-aggregates_part1.sql.out (hunk 9 above)
Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/udf/postgreSQL/udf-aggregates_part1.sql.out]: It seems `bool_and/or` has the similar issue;
```
scala> sql("select bool_and(a) from values (true), (true), (false) t(a)").explain(true)
== Analyzed Logical Plan ==
bool_and(a): boolean
Aggregate [bool_and(bool_and, a#12) AS bool_and(a)#14]
           ^^^^^^^^^^^^^^^^^^^
+- SubqueryAlias `t`
   +- LocalRelation [a#12]
```

**Comment by maropu** [comment]
@amanomer Yea, I think so. Can you address them?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 10]
```diff
@@ -635,7 +635,6 @@ object FunctionRegistry {
       (implicit tag: ClassTag[T]): (String, (ExpressionInfo, FunctionBuilder)) = {
     val constructors = tag.runtimeClass.getConstructors
       .filter(_.getParameterTypes.head == classOf[String])
-    assert(constructors.length == 1)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: `assert(constructors.length >= 1)`? cuz these classes must have the name field in the head?

Comment (amanomer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: `assert(constructors.length >= 1, s"there is no constructor for ${tag.runtimeClass} which takes String as first argument")`

cc @maropu

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: sounds good

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: +1, thanks!

**Review by maropu — APPROVED** [review]

**Comment by amanomer** [comment]
```
**Error Message**

java.sql.SQLException: Could not open client transport with JDBC Uri: jdbc:hive2://localhost:18954: java.net.ConnectException: Connection refused (Connection refused)
```
I think this failure is not related

**Comment by cloud-fan** [comment]
retest this please

**Comment by amanomer** [comment]
@cloud-fan @maropu 
Do we need to use `expressionWithAlias` for Average and ApproximatePercentile, too? In this PR?

**Comment by cloud-fan** [comment]
yes, please

**Comment by amanomer** [comment]
Tests have passed. Kindly review

cc @cloud-fan @maropu

**Comment by amanomer** [comment]
Tests have passed. Kindly review.
 @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala [hunk 11]
```diff
@@ -67,19 +67,25 @@ import org.apache.spark.sql.types._
...
     child: Expression,
     percentageExpression: Expression,
     accuracyExpression: Expression,
     override val mutableAggBufferOffset: Int,
     override val inputAggBufferOffset: Int)
   extends TypedImperativeAggregate[PercentileDigest] with ImplicitCastInputTypes {
 
-  def this(child: Expression, percentageExpression: Expression, accuracyExpression: Expression) = {
-    this(child, percentageExpression, accuracyExpression, 0, 0)
+  def this(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: who calls this constructor?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala [hunk 12]
```diff
@@ -67,19 +67,25 @@ import org.apache.spark.sql.types._
...
+    funcName: String,
+    child: Expression,
+    percentageExpression: Expression,
+    accuracyExpression: Expression) = {
+    this(funcName, child, percentageExpression, accuracyExpression, 0, 0)
   }
 
-  def this(child: Expression, percentageExpression: Expression) = {
-    this(child, percentageExpression, Literal(ApproximatePercentile.DEFAULT_PERCENTILE_ACCURACY))
+  def this(funcName: String, child: Expression, percentageExpression: Expression) = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: ditto

**Comment by amanomer** [comment]
@cloud-fan @maropu kindly review this PR

**Review by maropu — APPROVED** [review]

**Comment by amanomer** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala [hunk 13]
```diff
@@ -321,4 +331,22 @@ object ApproximatePercentile {
   }
 
   val serializer: PercentileDigestSerializer = new PercentileDigestSerializer
+
+  def apply(
+    child: Expression,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: nit: 4 space indentation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala [hunk 14]
```diff
@@ -67,19 +67,25 @@ import org.apache.spark.sql.types._
...
     percentageExpression: Expression,
     accuracyExpression: Expression,
     override val mutableAggBufferOffset: Int,
     override val inputAggBufferOffset: Int)
   extends TypedImperativeAggregate[PercentileDigest] with ImplicitCastInputTypes {
 
-  def this(child: Expression, percentageExpression: Expression, accuracyExpression: Expression) = {
-    this(child, percentageExpression, accuracyExpression, 0, 0)
+  def this(
+    funcName: String,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala [hunk 15]
```diff
@@ -35,9 +35,11 @@ import org.apache.spark.sql.types._
        -3 days -11 hours -59 minutes -59 seconds
   """,
   since = "1.0.0")
-case class Average(child: Expression) extends DeclarativeAggregate with ImplicitCastInputTypes {
+case class Average(
+      funcName: String, child: Expression)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala [hunk 16]
```diff
@@ -321,4 +331,22 @@ object ApproximatePercentile {
...
+  def apply(
+    child: Expression,
+    percentageExpression: Expression,
+    accuracyExpression: Expression,
+    mutableAggBufferOffset: Int, inputAggBufferOffset: Int): ApproximatePercentile =
+    new ApproximatePercentile("percentile_approx", child, percentageExpression,
+      accuracyExpression, mutableAggBufferOffset, inputAggBufferOffset)
+
+  def apply(
+    child: Expression,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: ditto

**Review by cloud-fan — APPROVED** [review]
LGTM except some code style issues.

**Comment by amanomer** [comment]
cc @cloud-fan @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala [hunk 17]
```diff
@@ -321,4 +331,22 @@ object ApproximatePercentile {
   }
 
   val serializer: PercentileDigestSerializer = new PercentileDigestSerializer
+
+  def apply(
+      child: Expression,
+      percentageExpression: Expression,
+      accuracyExpression: Expression,
+      mutableAggBufferOffset: Int, inputAggBufferOffset: Int): ApproximatePercentile =
+    new ApproximatePercentile("percentile_approx", child, percentageExpression,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: sorry to bring this up so late, but after more thoughts, I feel it's too invasive to add a `funcName` parameter to expression constructor: we need to add a lot of `apply` methods and constructors to provide the default function name.

This problem wasn't obvious when we were changing `BoolAnd`, but it's a bit obvious now for `ApproximatePercentile` which already have a lot of constructors.

I'd like to propose another idea: using mutable states via `TreeNodeTag`. The idea is pretty simple: an expression can have an optional alias, which is stored in the expression itself via `TreeNodeTag`. We can update `ApproximatePercentile.prettyName` to `getTagValue(THE_TAG).getOrElse("percentile_approx")`.

In `expressionWithAlias`, it can just call `expression`, and set the TreeNodeTag with the alias at the end.

Comment (amanomer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: It's a great idea. I will implement and update this PR. Thanks @cloud-fan

**Comment by amanomer** [comment]
There is duplicate code in expression and expressionWithAlias.

**Comment by amanomer** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 18]
```diff
@@ -573,7 +573,7 @@ object FunctionRegistry {
   val functionSet: Set[FunctionIdentifier] = builtin.listFunction().toSet
 
   /** See usage above. */
-  private def expression[T <: Expression](name: String)
+  private def expression[T <: Expression](name: String, isAliasName: Boolean = false)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: nit: `setAlias: Boolean = false`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 19]
```diff
@@ -286,6 +286,8 @@ abstract class Expression extends TreeNode[Expression] {
   override def simpleStringWithNodeId(): String = {
     throw new UnsupportedOperationException(s"$nodeName does not implement simpleStringWithNodeId")
   }
+
+  val FUNC_ALIAS = TreeNodeTag[String]("functionAliasName")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: We can define it in `object FunctionRegistry`. This is kind of a map key, and we don't need to create a different instance for each expression.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 20]
```diff
@@ -282,34 +282,34 @@ object FunctionRegistry {
 
     // aggregate functions
     expression[HyperLogLogPlusPlus]("approx_count_distinct"),
-    expression[Average]("avg"),
+    expressionWithAlias[Average]("avg"),
+    expressionWithAlias[Average]("mean"),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Seems we don't need `expressionWithAlias` at all, just call  `expression[Average]("mean", setAlias = true)`.

And `expression[Average]("avg")` can remain unchanged, as `avg` is already the default name.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: And we don't need to reorder them now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala [hunk 21]
```diff
@@ -185,7 +185,7 @@ case class ApproximatePercentile(
     if (returnPercentileArray) ArrayType(child.dataType, false) else child.dataType
   }
 
-  override def prettyName: String = "percentile_approx"
+  override def nodeName: String = getTagValue(FUNC_ALIAS).getOrElse("percentile_approx")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: let's override `prettyName` as before.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/First.scala [hunk 22]
```diff
@@ -113,5 +113,7 @@ case class First(child: Expression, ignoreNullsExpr: Expression)
 
   override lazy val evaluateExpression: AttributeReference = first
 
-  override def toString: String = s"first($child)${if (ignoreNulls) " ignore nulls"}"
+  override def nodeName: String = getTagValue(FUNC_ALIAS).getOrElse("first")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/First.scala]: ditto, override `prettyName` seems better.

**Comment by cloud-fan** [comment]
in general looks good, thanks for updating!

**Comment by amanomer** [comment]
cc @cloud-fan

**Comment by amanomer** [comment]
retest this

**Review by cloud-fan — APPROVED** [review]

**Comment by amanomer** [comment]
Tests have passed. cc @cloud-fan

**Comment by cloud-fan** [comment]
thanks, merging to master!
